src/accumulator/proof.rs

```diff
@@ -0,0 +1,293 @@
+    ///   // 14                                                         
+    ///   // |-----------------\                               
+    ///   // 12                13                         
+    ///   // |---------\       |--------\               
+    ///   // 08       09       10       11         
+    ///   // |----\   |----\   |----\   |----\       
+    ///   // 00   01  02   03  04   05  06   07
+    ///   // For proving 0, we need 01, 09 and 13's hashes. 00, 08, 12 and 14 can be calculated
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a").unwrap());
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b").unwrap());
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("29590a14c1b09384b94a2c0e94bf821ca75b62eacebc47893397ca88e3bbcbd7").unwrap());
+    ///   
+    ///   let mut hashes = Vec::new();
+    ///   for i in test_values {
+    ///     let mut engine = bitcoin_hashes::sha256::Hash::engine();
+    ///      engine.input(&[i]);
+    ///      let hash = sha256::Hash::from_engine(engine);
+    ///     hashes.push(hash);
+    ///   }
+    ///
+    ///   s.modify(&hashes, &vec![]);
+    ///   let p = Proof::new(targets, proof_hashes);
+    ///   assert!(p.verify(&vec![hashes[0]] , &s).expect("No error should happens"));
+    ///```
+    pub fn verify(&self, del_hashes: &Vec<sha256::Hash>, stump: &Stump) -> Result<bool, String> {
+        let calculated_roots = self.create_root_candidates(del_hashes, stump)?;
+        let stump_roots = util::num_roots(stump.leafs);
+        let mut number_matched_roots = 0;
+
+        for i in 0..stump_roots {
+            if calculated_roots[number_matched_roots] == stump.roots[stump.roots.len() - (i+1)]
+            {
+                number_matched_roots += 1;
+            }
+        }
+
+        if calculated_roots.len() != number_matched_roots && calculated_roots.len() != 0 {
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    pub fn create_root_candidates(
+        &self,
+        del_hashes: &Vec<sha256::Hash>,
+        stump: &Stump,
+    ) -> Result<Vec<sha256::Hash>, String> {
+        // Where all the root hashes that we've calculated will go to.
+        let total_rows = util::tree_rows(stump.leafs);
+        let mut targets = self.targets.clone();
+
+        // Where all the parent hashes we've calculated in a given row will go to.
+        let mut calculated_root_hashes =
+            Vec::<sha256::Hash>::with_capacity(util::num_roots(stump.leafs) as usize);
+
+        // As we calculate nodes upwards, it accumulates here
+        let mut nodes = Proof::make_hash_and_pos(del_hashes, &targets);
+        let mut row = 0;
+        let mut proof_hashes_index = 0 as usize;
+
+        while row <= total_rows {
+            let res = util::detect_row_hashes(&targets, row, total_rows);
```

Comment:
I *think* the loop here is incorrect because there's no sorted merge. Maybe I'm wrong but the rationale I had for sorting is something like this:

```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
00  01          04  05      
```

When you do

```
targets.push(next_to_prove);
nodes.push(HashAndPos::mask(hash, next_to_prove);
```

the nodes now become out of order. Say if we're proving `[00, 09, 11]` with the proof being `[01, 10]`:

`00` gets hashed with proof of `01` and the code would append `08` and `nodes` become: `[00, 09, 11, 08]`.
`09` wouldn't be able to hash up to `12` because the sibling check at line 169 would fail as `11` is not a sibling of `09`.

Go playground code: https://go.dev/play/p/W6ujymRI63G

Reply:
> You mean that when adding a new target, I should sort the vector before continuing?

Basically yeah. The loop that I had in the Go code only works when the nodes are sorted. I think the bug only shows up when the leaves are not on the same row.

>  I think I should add more tests for this kind of tree, where leafs are not at the same row.

Yeah that'd be good. You could add `Roots` to the json testcases so that you don't have to generate the roots from the leaves during the tests. I don't think the current setup works since there's no deletions implemented yet.